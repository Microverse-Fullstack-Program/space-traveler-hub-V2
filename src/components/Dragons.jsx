import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchDragons } from '../redux/dragons/dragonsSlice';
import DragonPage from './DragonsPage';

const Dragons = () => {
  const dispatch = useDispatch();
  const dragons = useSelector((state) => state.dragons.dragons);
  const error = useSelector((state) => state.dragons.error);
  const isLoading = useSelector((state) => state.dragons.isLoading);

  useEffect(() => {
    if (dragons.length === 0) dispatch(fetchDragons());
  }, [dispatch, dragons.length]);

  return (
    <div className="dragon-list">
      {error && <p>Error Loading dragons</p>}
      {dragons
        && dragons.map((dragon) => <DragonPage key={dragon.id} dragon={dragon} />)}
      {isLoading && <p>Loading...</p>}
    </div>
  );
};

export default Dragons;
