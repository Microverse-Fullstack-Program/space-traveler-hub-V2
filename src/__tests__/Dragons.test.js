import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderWithProviders from '../utils/test-utils';
import Dragons from '../components/Dragons';
import setupStore from '../redux/store';
import server from '../mocks/server';

describe('Dragon snapshot', () => {
  test('should render heading correctly', () => {
    renderWithProviders(
      <Provider store={setupStore({})}>
        <Dragons />
      </Provider>,
    );
    const dragons = document.querySelector('.dragon-list');
    expect(dragons).toMatchSnapshot();
  });
});

beforeAll(() => server.listen({ onUnhandledRequest: 'bypass' })); // Enable the mocking in tests.
afterEach(() => server.resetHandlers()); // Reset any runtime handlers tests may use.
afterAll(() => server.close()); // Clean up once the tests are done.

describe('Test dragons before fetching API', () => {
  test('Test for Loading... state', () => {
    renderWithProviders(<Dragons />);
    expect(screen.queryByText(/Loading.../i)).toBeInTheDocument();
    expect(screen.queryByText(/Reserve Dragon/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Dragon 1/i)).not.toBeInTheDocument();
  });
});

describe('Test dragons after fetching API', () => {
  test('Test if dragons are rendered after fetching', async () => {
    renderWithProviders(<Dragons />);
    expect(await screen.findAllByText(/Reserve Dragon/i)).toHaveLength(2);
    expect(await screen.findAllByText(/Dragon 1/i)).toHaveLength(1);
    expect(screen.queryByText(/Loading.../i)).not.toBeInTheDocument();
  });
});
