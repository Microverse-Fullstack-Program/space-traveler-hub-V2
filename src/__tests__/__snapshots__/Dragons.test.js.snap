// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragon snapshot should render heading correctly 1`] = `
<div
  class="dragon-list"
>
  <p>
    Loading...
  </p>
</div>
`;
